### PR TITLE
Add cluster ip for Kubernetes Nat Manager

### PR DESCRIPTION
**Pull request by matkt**
Signed-off-by: Karim TAAM <karim.t2am@gmail.com>

## PR description

Kubernetes nat manager could work with `LoadBalancer` services. This PR add compatibility with `ClusterIP` services. 
A code reactor was made in order to facilitate the addition of new services thereafter such as the `NodePort`

Sample of `ClusterIP` service :

```yaml
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app.kubernetes.io/name: besu
    app.kubernetes.io/release: "1.0.0"
  name: besu
spec:
  ports:
  - name: "json-rpc"
    port: 8546
    targetPort: 8545
  - name: "rlpx"
    port: 30303
    targetPort: 30303
  selector:
    app.kubernetes.io/name: besu
    app.kubernetes.io/release: "1.0.0"
  type: ClusterIP

```

You can find the IP that will be used by BESU by launching the following command

> kubectl describe services besu  

```             
Name:              besu
Namespace:         default
Labels:            app.kubernetes.io/name=besu
                   app.kubernetes.io/release=1.0.0
Annotations:       kubectl.kubernetes.io/last-applied-configuration:
                     {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"labels":{"app.kubernetes.io/name":"besu","app.kubernetes.io/release":"1....
Selector:          app.kubernetes.io/name=besu,app.kubernetes.io/release=1.0.0
Type:              ClusterIP
IP:                **<IP USED BY KUBERNETES NAT MANAGER>**
Port:              json-rpc  8546/TCP
TargetPort:        8545/TCP
Endpoints:         <none>
Port:              rlpx  30303/TCP
TargetPort:        30303/TCP
Endpoints:         <none>
Session Affinity:  None
Events:            <none>
```

### Update of the fallback mechanism

- If the `nat-method` is equal to `AUTO` (default). A bad configuration (invalid service, ingress not found, etc.) will cause a switch to the `NONE` mode.

- If the `nat-method` is equal to `KUBERNETES` and `Xnat-method-fallback-enabled` is equal to `true`. A bad configuration will cause a switch to the `NONE` mode.

- If the `nat-method` is equal to `KUBERNETES` and `Xnat-method-fallback-enabled`is equal to `false`. A bad configuration will cause a stop of `BESU` in order to indicate to him that the automatic configuration has failed

### New error message added

If you are trying to use not implemented service you will now have an issue :
> Failed update information using pod metadata : NodePort is not implemented.

### Tested

| Test | Result expected  |   | 
|--|---|---|
|     Deploy with a ClusterIP      |  Work | OK  |  
|     Deploy with a LoadBalancer      |  Work  | OK  |  
|     Deploy with a NodePort   (AUTO)    |  Throw an exception and switch to NONE | OK  | 
|      Deploy with a Missing service  (AUTO)   |  Throw an exception and switch to NONE | OK  |  
|       Deploy without fallback (KUBERNETES  and fallback=false)    | Stop besu with an exception  | OK  |  
|       Deploy without fallback (KUBERNETES  and fallback=true)    |  Throw an exception and switch to NONE  | OK  |

### CODE REVIEW

File: nat/src/main/java/org/hyperledger/besu/nat/NatService.java [hunk 1]
```diff
@@ -88,10 +96,14 @@ public void start() {
         getNatManager().orElseThrow().start();
       } catch (Exception e) {
         LOG.debug(
-            "Nat manager failed to configure itself automatically due to the following reason "
-                + e.getMessage()
-                + ". NONE mode will be used");
-        disableNatManager();
+            String.format(
```
Comment (abdelhamidbakhta) [inline nat/src/main/java/org/hyperledger/besu/nat/NatService.java:99]: Don't use `String.format`, use standard logging placeholder instead.

File: nat/src/main/java/org/hyperledger/besu/nat/kubernetes/service/ClusterIpBasedDetector.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.hyperledger.besu.nat.kubernetes.service;
+
+import org.hyperledger.besu.nat.core.IpDetector;
+
+import java.util.Optional;
+
+import io.kubernetes.client.models.V1Service;
+
+public class ClusterIpBasedDetector implements IpDetector {
```
Comment (abdelhamidbakhta) [inline nat/src/main/java/org/hyperledger/besu/nat/kubernetes/service/ClusterIpBasedDetector.java:23]: Not worth creating a class for such a simple implementation IMO. You can inject the code as a lambda directly.

File: nat/src/main/java/org/hyperledger/besu/nat/kubernetes/KubernetesNatManager.java [hunk 3]
```diff
@@ -152,4 +144,16 @@ protected void doStop() {
   public CompletableFuture<List<NatPortMapping>> getPortMappings() {
     return CompletableFuture.completedFuture(forwardedPorts);
   }
+
+  private IpDetector getIpDetector(final V1Service v1Service) throws NatInitializationException {
+    final String serviceType = v1Service.getSpec().getType();
+    switch (KubernetesServiceType.fromName(serviceType)) {
+      case CLUSTER_IP:
+        return new ClusterIpBasedDetector(v1Service);
```
Comment (abdelhamidbakhta) [inline nat/src/main/java/org/hyperledger/besu/nat/kubernetes/KubernetesNatManager.java:152]: Return a lambda instead IMO

File: nat/src/test/java/org/hyperledger/besu/nat/NatServiceTest.java [hunk 4]
```diff
@@ -197,6 +197,24 @@ public void assertThatManagerSwitchToNoneForInvalidNatEnvironment()
     assertThat(natService.queryLocalIPAddress(fallbackLocalIp)).isEqualTo(fallbackLocalIp);
   }
 
+  @Test(expected = RuntimeException.class)
```
Comment (abdelhamidbakhta) [inline nat/src/test/java/org/hyperledger/besu/nat/NatServiceTest.java:200]: Use `assertThatThrownBy` instead.
Example here: https://github.com/hyperledger/besu/blob/71f2872ecaaca05b6f70276589bee69bed51cd18/besu/src/test/java/org/hyperledger/besu/chainexport/RlpBlockExporterTest.java#L213

File: nat/src/test/java/org/hyperledger/besu/nat/docker/DockerNatManagerTest.java [hunk 5]
```diff
@@ -61,7 +63,14 @@ public void assertThatExternalIPIsEqualToRemoteHost()
   @Test
   public void assertThatExternalIPIsEqualToDefaultHostIfIpDetectorCannotRetrieveIP()
       throws ExecutionException, InterruptedException {
-    when(hostBasedIpDetector.detectExternalIp()).thenReturn(Optional.empty());
+    NatManager natManager =
```
Comment (abdelhamidbakhta) [inline nat/src/test/java/org/hyperledger/besu/nat/docker/DockerNatManagerTest.java:66]: final ?

File: nat/src/main/java/org/hyperledger/besu/nat/NatService.java (hunk 1 above)
Comment (matkt) [reply nat/src/main/java/org/hyperledger/besu/nat/NatService.java:99]: Done

File: nat/src/main/java/org/hyperledger/besu/nat/kubernetes/service/ClusterIpBasedDetector.java (hunk 2 above)
Comment (matkt) [reply nat/src/main/java/org/hyperledger/besu/nat/kubernetes/service/ClusterIpBasedDetector.java:23]: Good catch

File: nat/src/main/java/org/hyperledger/besu/nat/kubernetes/KubernetesNatManager.java (hunk 3 above)
Comment (matkt) [reply nat/src/main/java/org/hyperledger/besu/nat/kubernetes/KubernetesNatManager.java:152]: Done

File: nat/src/test/java/org/hyperledger/besu/nat/NatServiceTest.java (hunk 4 above)
Comment (matkt) [reply nat/src/test/java/org/hyperledger/besu/nat/NatServiceTest.java]: Updated

File: nat/src/test/java/org/hyperledger/besu/nat/docker/DockerNatManagerTest.java (hunk 5 above)
Comment (matkt) [reply nat/src/test/java/org/hyperledger/besu/nat/docker/DockerNatManagerTest.java]: You right . Done

**Pull request merged**